### ISSUE TITLE
Global store discussion

### ISSUE THREAD

**Issue by davestewart** [davestewart/vue-class-store#13]
Starting this thread to prevent #12 from getting cluttered

**Comment by davestewart** [comment]
It may be too early to declare victory, but I appear to have gotten a nice TypeScript workflow going using:

- a `RootStore` class instance
- a global `Vue.use()`
- and [Module Augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) which provides autocomplete throughout the project

The root store class looks like this:

```ts
export class RootStore {
  $add (name: string, instance) {
    if (name.startsWith('$')) {
      throw new Error('The "$" character is reserved for store instance methods')
    }
    this[name] = instance
  }

  $remove (name: string) {
    delete this[name]
  }

  constructor (modules: Record<string, any> = {}) {
    Object.keys(modules).forEach(key => {
      this.$add(key, modules[key])
    })
  }
}
```

install looks like this:

```ts
export function install (Vue: VueConstructor, modules: Record<string, any> = {}) {
  Vue.prototype.$store = new RootStore(modules)
}
```

My store index looks like this:

```ts
import Vue from 'vue'
import { RootStore, install } from '@/packages/store'

// local modules
import app, { AppStore } from './app'
import ui, { UiStore } from './ui'

// tell typescript about modules
declare module '@/packages/store' {
  interface RootStore {
    app: AppStore,
    ui: UiStore,
  }
}

// tell vue about the built-in store
declare module 'vue/types/vue' {
  interface Vue {
    $store: RootStore
  }
}

// install and add modules
Vue.use(install, {
  app,
  ui,
})

// ensure other modules importing store directly know what it is
export default Vue.prototype.$store as RootStore
```

You can add modules individually using a combination of instantiation and augmentation:

```ts
// import root store
import store from '@/store'

// augment root store with the new property 
declare module '@/packages/store' {
  interface RootStore {
    tools: ToolsStore,
  }
}

// local vue class store setup
@VueStore
class ToolsStore {
  resize () {
    bus.call('tools/windows/resize')
  }
}

// physically add the local store to the root store
store.$add('tools', new ToolsStore())
```

Autocomplete works great in Vue files:

![image](https://user-images.githubusercontent.com/132681/102559447-ce600180-40c7-11eb-9f36-8968ee9e2000.png)

I know some of the TS boilerplate looks a little funky, but:

- it allows a root store to be created without explicitly declaring properties
- it allows local stores to be added independently of the root store declarations

I have a minor issue with the `RootStore::$add` not being recognised in one file without me adding the signature to the root augmentation. I don't know why this is (but I'm not an expert in this). Once that is done; everything plays ball.

Happy to hear any thoughts!

**Comment by Mootook** [comment]
@davestewart -- Took me a bit as I'm not as familiar with TypeScript, but I think I get the gist. Looks nice. Would this replace the entire provide/inject usage of the current release?
```js
setup () {
  const { tools } = useStore()
  return { tool }
  // vs. 
  return { tools: this.$store.tools }
}
```
Globally attaching the store would be great and eliminate some boilerplate at the component level. Also, just a note, Vue is no longer exposed in the vue3 package. It would instead have to be attached to the app instance...something like:
```js
app.config.globalProperties.$store = new RootStore(modules) // or something
```

**Comment by davestewart** [comment]
Hmmm. We have inverse knowledge-sets!

I can only really comment regarding Vue 2 right now, but I'm finding that `this.$store.<model>` just feels a little "more global" than `this.$model`, but you can do both, of course.

Vue 3 wise, I don't suppose it makes any difference, but you would have more choice.

Seeing as Vue 3's setup function does make it easier to type variables, perhaps it's not even needed in Vue 3?

**Comment by Mootook** [comment]
I agree with `this.$store.anything` feeling more global. And it does steer clear of the risk of assigning some variable to a nested property or destructuring an object and losing reactivity. I know even vue3 has some guidelines on that especially regarding props in the composition API. 

Though, I have found a nice workflow with the current vue-class-store partnered with a jsconfig and jsdoc, which allows for typing "guides" but not necessarily strict type checking like what TS offers.

<img width="288" alt="Screen Shot 2020-12-18 at 9 41 29 AM" src="https://user-images.githubusercontent.com/42306342/102644171-37219b00-4115-11eb-88e3-d7dd00b8e62b.png">

<img width="442" alt="Screen Shot 2020-12-18 at 9 41 16 AM" src="https://user-images.githubusercontent.com/42306342/102644175-3852c800-4115-11eb-8b18-7366b23db041.png">

I think the way you posted works really well and keeps in tune with how a lot of things are packaged into Vue2 apps. So perhaps the best idea is just to press on with that. Vue3 might have to be its own discussion.
